policyeval/broker.go

```diff
@@ -0,0 +1,347 @@
+	logger := b.logger.With("eval_id", evalID, "token", token)
+
+	logger.Debug("ack eval", "eval_id", evalID, "token", token)
+
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	// Lookup the unack'd eval.
+	unack, ok := b.unack[evalID]
+	if !ok {
+		return fmt.Errorf("evaluation ID not found")
+	}
+	if unack.Token != token {
+		return fmt.Errorf("token does not match for evaluation ID")
+	}
+
+	// Ensure we were able to stop the timer.
+	if !unack.NackTimer.Stop() {
+		return fmt.Errorf("evaluation ID Ack'd after Nack timer expiration")
+	}
+
+	// Cleanup.
+	delete(b.unack, evalID)
+	delete(b.enqueuedEvals, evalID)
+	delete(b.enqueuedPolicies, unack.Eval.Policy.ID)
+
+	b.logger.Debug("eval ack'd", "policy_id", unack.Eval.Policy.ID)
+	return nil
+}
+
+// Nack is used to mark an eval as not completed.
+func (b *Broker) Nack(evalID, token string) error {
+	logger := b.logger.With("eval_id", evalID, "token", token)
+
+	logger.Debug("nack eval")
+
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	// Lookup the unack'd eval
+	unack, ok := b.unack[evalID]
+	if !ok {
+		return fmt.Errorf("evaluation ID not found")
+	}
+	if unack.Token != token {
+		return fmt.Errorf("token does not match for evaluation ID")
+	}
+
+	logger = logger.With("policy_id", unack.Eval.Policy.ID)
+
+	// Stop the timer, doesn't matter if we've missed it.
+	unack.NackTimer.Stop()
+
+	// Cleanup.
+	delete(b.unack, evalID)
+
+	// Check if we've hit the delivery limit.
+	if dequeues := b.enqueuedEvals[evalID]; dequeues >= b.deliveryLimit {
+		logger.Warn("eval delivery limit reached", "count", dequeues, "limit", b.deliveryLimit)
+		return nil
+	}
+
+	// Re-enqueue eval to try again.
+	b.enqueueLocked(unack.Eval, token)
+	logger.Info("eval nack'd, retrying it")
+	return nil
+}
+
+// PendingEvaluations is a list of waiting evaluations.
+// We implement the container/heap interface so that this is a
+// priority queue
+type PendingEvaluations []*sdk.ScalingEvaluation
+
+// Len is for the sorting interface
+func (p PendingEvaluations) Len() int {
+	return len(p)
+}
+
+// Less is for the sorting interface. We flip the check
+// so that the "min" in the min-heap is the element with the
+// highest priority
+func (p PendingEvaluations) Less(i, j int) bool {
+	if p[i].Policy.Priority != p[j].Policy.Priority {
+		return !(p[i].Policy.Priority < p[j].Policy.Priority)
+	}
+	return p[i].CreateTime.Before(p[j].CreateTime)
+}
+
+// Swap is for the sorting interface
+func (p PendingEvaluations) Swap(i, j int) {
+	p[i], p[j] = p[j], p[i]
+}
+
+// Push is used to add a new evaluation to the slice
+func (p *PendingEvaluations) Push(e interface{}) {
+	*p = append(*p, e.(*sdk.ScalingEvaluation))
+}
+
+// Pop is used to remove an evaluation from the slice
+func (p *PendingEvaluations) Pop() interface{} {
+	n := len(*p)
+	e := (*p)[n-1]
+	(*p)[n-1] = nil
+	*p = (*p)[:n-1]
+	return e
+}
+
+// Peek is used to peek at the next element that would be popped
+func (p PendingEvaluations) Peek() *sdk.ScalingEvaluation {
```

Comment:
it doesn't appear that we use `Peek()` here

Reply:
Ah right. In Nomad this is used by workers that can handle multiple queues, but that's not the case for use so I will remove it.